internal/controllers/machine/machine_controller_kubelet_serving_csr.go

```diff
@@ -0,0 +1,173 @@
+package machine
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// KubeletServingSignerName defines the name for the kubelet serving certificate signer
+	KubeletServingSignerName = "kubernetes.io/kubelet-serving"
+
+	// NodesPrefix defines the prefix name for a node.
+	NodesPrefix = "system:node:"
+
+	// NodesGroup defines the group name for a node.
+	NodesGroup = "system:nodes"
+)
+
+func (r *Reconciler) reconcileKubletServingCertificateSigningRequest(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx, "machine", machine.Name, "namespace", machine.Namespace)
+	log = log.WithValues("cluster", cluster.Name)
+
+	// TODO can NodeRef be nil at this point?
+	nodeRef := machine.Status.NodeRef
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	restConfig, err := remote.RESTConfig(ctx, controllerName, r.Client, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+
+	node := corev1.Node{}
+	if err := remoteClient.Get(ctx, client.ObjectKey{Name: nodeRef.Name}, &node); err != nil {
+		return ctrl.Result{}, err
```

Comment:
this logic in general could be modelled against meaningful abstractions of the domain, e.g getServingCert(), getClientCert(), validateServingCert() which can then unit test and change impl transparently.